cpp/tests/experimental/range/multipolygon_range_test.cu

```diff
@@ -0,0 +1,340 @@
+
+    thrust::copy(rmm::exec_policy(stream()),
+                 rng.multipolygon_segment_count_begin(),
+                 rng.multipolygon_segment_count_end(),
+                 got.begin());
+
+    auto d_expected = thrust::device_vector<std::size_t>(expected_segment_counts.begin(),
+                                                         expected_segment_counts.end());
+
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(got, d_expected);
+  }
+};
+
+TYPED_TEST_CASE(MultipolygonRangeTest, FloatingPointTypes);
+
+TYPED_TEST(MultipolygonRangeTest, SegmentIterators)
+{
+  using T = TypeParam;
+  using P = vec_2d<T>;
+  using S = segment<T>;
+  CUSPATIAL_RUN_TEST(this->run_multipolygon_segment_iterator_single,
+                     {0, 1},
+                     {0, 1},
+                     {0, 4},
+                     {{0, 0}, {1, 0}, {1, 1}, {0, 0}},
+                     {S{P{0, 0}, P{1, 0}}, S{P{1, 0}, P{1, 1}}, S{P{1, 1}, P{0, 0}}});
+}
+
+TYPED_TEST(MultipolygonRangeTest, SegmentIterators2)
+{
+  CUSPATIAL_RUN_TEST(this->run_multipolygon_segment_iterator_single,
+                     {0, 1},
+                     {0, 2},
+                     {0, 4, 8},
+                     {{0, 0}, {1, 0}, {1, 1}, {0, 0}, {10, 10}, {11, 10}, {11, 11}, {10, 10}},
+                     {{{0, 0}, {1, 0}},
+                      {{1, 0}, {1, 1}},
+                      {{1, 1}, {0, 0}},
+                      {{10, 10}, {11, 10}},
+                      {{11, 10}, {11, 11}},
+                      {{11, 11}, {10, 10}}});
+}
+
+TYPED_TEST(MultipolygonRangeTest, SegmentIterators3)
+{
+  CUSPATIAL_RUN_TEST(this->run_multipolygon_segment_iterator_single,
+                     {0, 2},
+                     {0, 1, 2},
+                     {0, 4, 8},
+                     {{0, 0}, {1, 0}, {1, 1}, {0, 0}, {10, 10}, {11, 10}, {11, 11}, {10, 10}},
+                     {{{0, 0}, {1, 0}},
+                      {{1, 0}, {1, 1}},
+                      {{1, 1}, {0, 0}},
+                      {{10, 10}, {11, 10}},
+                      {{11, 10}, {11, 11}},
+                      {{11, 11}, {10, 10}}});
+}
+
+TYPED_TEST(MultipolygonRangeTest, SegmentIterators4)
+{
+  CUSPATIAL_RUN_TEST(this->run_multipolygon_segment_iterator_single,
+                     {0, 1, 2},
+                     {0, 1, 2},
+                     {0, 4, 8},
+                     {{0, 0}, {1, 0}, {1, 1}, {0, 0}, {10, 10}, {11, 10}, {11, 11}, {10, 10}},
+                     {{{0, 0}, {1, 0}},
+                      {{1, 0}, {1, 1}},
+                      {{1, 1}, {0, 0}},
+                      {{10, 10}, {11, 10}},
+                      {{11, 10}, {11, 11}},
+                      {{11, 11}, {10, 10}}});
+}
+
+TYPED_TEST(MultipolygonRangeTest, PerMultipolygonCountIterator)
+{
+  CUSPATIAL_RUN_TEST(this->run_multipolygon_point_count_iterator_single,
+                     {0, 1},
+                     {0, 1},
+                     {0, 4},
+                     {{0, 0}, {1, 0}, {1, 1}, {0, 1}},
```

Comment:
Point count iterators for multipolygons seems a bit wrong... This isn't a valid test to start with, the ring should be closed. And the count of point should be the the difference between offset -1. Because the first and the last point overlaps.

Reply:
After thinking on this again, I think the point count iterator should be ok to return just the difference between the offsets. The iterator didn't say anything about "unique point counts", but just "point counts". Besides, this allows the segment count iterator to treat the point count iterator the same way for ring and non-ring linestrings, reducing code complexity.

Test case setup is still wrong though. Needs to fix.